### PR TITLE
[kore-rpc] Initial working implementation + test harness stub 

### PR DESCRIPTION
**Pull request by goodlyrottenapple**
This initial implementation sends back dummy values for all the routes. We first need to address #3143 and #3144 to provide correct behavior.

This PR also includes a stub test harness, which is a Python script that launches a new instance of `kore-rpc` and tests the output, generated per the given RPC JSON request, against the golden response. It can be run via `nix build .#test-rpc` and should eventually be added to the CI test run.

There is also a document describing the current JSON RPC API:

https://github.com/runtimeverification/haskell-backend/blob/eea0670eac043684ad8d0ad812eb584d0437007a/docs/2022-07-18-JSON-RPC-Server-API.md

### CODE REVIEW

File: kore/src/Kore/Syntax/Json.hs
Comment (goodlyrottenapple) [inline kore/src/Kore/Syntax/Json.hs:18]: @jberthold I'm re-exporting some of the internals here. Hope that's ok?

Comment (jberthold) [reply kore/src/Kore/Syntax/Json.hs:18]: No problem at all, makes sense to export the conversion functions. I would maybe 
* export `KoreJson(..)` rather than `term` 
* also export `KorePattern` (type only, without constructors)
but fine either way

File: kore/app/rpc/Main.hs
Comment (goodlyrottenapple) [reply kore/app/rpc/Main.hs:147]: sure

File: kore/app/rpc/Main.hs
Comment (goodlyrottenapple) [reply kore/app/rpc/Main.hs:166]: I've added comments describing whats going on

File: kore/app/rpc/Main.hs
Comment (goodlyrottenapple) [reply kore/app/rpc/Main.hs:73]: I'm not sure... I've removed it

**Review by jberthold — COMMENTED** [review]
LGTM, except two small details to adjust

File: kore/app/rpc/Main.hs
Comment (goodlyrottenapple) [reply kore/app/rpc/Main.hs:89]: not sure what that range is tbh.

File: kore/src/Kore/JsonRpc.hs
Comment (jberthold) [reply kore/src/Kore/JsonRpc.hs:166]: And you are saying it should indeed be `contition` with a `t` in the middle? :shrug:

File: kore/app/rpc/Main.hs
Comment (jberthold) [reply kore/app/rpc/Main.hs:89]: `[1 .. 65535]` (but using [1..1024] require `root` privileges). This is very minor, though, illegal port numbers will crash when starting inside `runrerver` otherwise

File: kore/src/Kore/JsonRpc.hs
Comment (goodlyrottenapple) [reply kore/src/Kore/JsonRpc.hs:166]: sorry didnt notice that :D

File: kore/app/rpc/Main.hs
Comment (goodlyrottenapple) [reply kore/app/rpc/Main.hs:89]: yeah, it might also crash if the port is taken

**Review by jberthold — APPROVED** [review]
Might need a manual rebase but LGTM
(code should probably be merged to unblock others).

File: test/rpc-server/runTests.py
Comment (tothtamas28) [inline test/rpc-server/runTests.py:18]: This might not be a good solution: https://stackoverflow.com/a/7337584.

Did you consider implementing the server using HTTP? Alternatively, if raw sockets are to be used then perhaps  the client can read the socket until it successfully parses the JSON (or runs into an error), but that's harder to implement.

File: kore/app/rpc/Main.hs
Comment (JKTKops) [inline kore/app/rpc/Main.hs:165]: Could this line be split up somehow? e.g. above the call to `clockSomethingIO` add a new binding for `setupWithMVar :: MVar ? -> SolverSetup`?

File: kore/src/Kore/Log/JsonRpc.hs
Comment (JKTKops) [inline kore/src/Kore/Log/JsonRpc.hs:43]: This definition is slightly dangerous - if `msg` is more than one line, then `--log-format=oneline` could end up printing one log across multiple lines, which could break log-trawling scripts.

That said we don't have many log-trawling scripts in use so it might not be a huge deal. But perhaps a warning comment?

File: test/rpc-server/runTests.py
Comment (goodlyrottenapple) [reply test/rpc-server/runTests.py:18]: This is just a test script, the pyk implementation should obviously make sure these things are correct. Using HTTP seems like an overkill given the RPC server is meant to be used locally only, as far as I understand. I think the vscode RPC server also just uses raw sockets without HTTP.

Comment (tothtamas28) [reply test/rpc-server/runTests.py:18]: I'm not against using TCP without HTTP, but in that case it would be much easier to implement clients if the server defined a super thin application layer at least, e.g.:

- The JSON payload is written in a single line, terminated by newline (my preference); or
- The first fixed number of bytes is an integer indicating the size of the rest of the payload in bytes.

Otherwise, we probably won't be able to use the JSON parser in the Python standard library (but maybe we could us [ijson](https://github.com/ICRAR/ijson) instead).

Comment (goodlyrottenapple) [reply test/rpc-server/runTests.py:18]: The server does indeed send a single line terminated by a newline (or at least should!)

Comment (tothtamas28) [reply test/rpc-server/runTests.py:18]: That's great, it simplifies things a lot. Please document it in `JSON-RPC-Server-API.md` so that it becomes part of the contract.

File: kore/src/Kore/Log/JsonRpc.hs
Comment (goodlyrottenapple) [reply kore/src/Kore/Log/JsonRpc.hs:43]: I've added `Text.replace "\n" " "`, reckon that should do the trick?

File: test/rpc-server/runTests.py
Comment (goodlyrottenapple) [reply test/rpc-server/runTests.py:18]: added to the doc now

**Pull request merged**
